File: sdk/src/scratch.ts
import {
  AccountMeta,
  Connection,
  PublicKey,
  TransactionInstruction,
  type Commitment,
} from "@solana/web3.js";

import { DEFAULT_IFX_PROGRAM_ID, indexCapForTapeLen } from "./constants";
import { immortalCloseAuthority } from "./immortal-close";
import { planRecordOffsets, recordByteLength } from "./tape-layout";
import { decodeFrameAccount, type DecodedFrame, framePda } from "./layout";
import {
  accountDataSliceBinding,
  binding,
  inferBindingTy,
  remapBindingAccountIndex,
} from "./binding";
import { expr } from "./expr";
import type { LetArgs, LetBinding, ValueType } from "./types";
import {
  scratchValue,
  tyForIfxTy,
  type Cond,
  type IfxTy,
  type ScratchValue,
  type TypedExpr,
} from "./typed";
import type { CpiBuildResult } from "./cpi";
import type { IfElseArgs } from "./types";
import {
  buildIxAssert,
  buildIxLet,
  buildIxResetFrame,
  createIxCloseFrame,
  createIxCreateFrame,
  createIxIfElse,
  createIxCpi,
  mergeIxOpts,
  type CreateIxCreateFrameParams,
  type IxOpts,
} from "./ix";
import {
  LetAccountInput,
  toLetAccountMeta,
} from "./let-account";
import { LetIxBuilder } from "./let-builder";

export type { CreateIxCreateFrameParams } from "./ix";

/** Params for {@link FrameScratch.planPublicFrame} — `close_authority` is set to {@link immortalCloseAuthority}. */
export type PlanPublicFrameParams = Omit<
  CreateIxCreateFrameParams,
  "closeAuthority"
>;

/** Options for {@link FrameScratch.fetchDecodedFrame}. */
export type FrameScratchReadOpts = {
  commitment?: Commitment;
};

export type {
  Cond,
  ExprInput,
  IfxTy,
  ScratchValue,
  TypedExpr,
} from "./typed";

/** Result of {@link FrameScratch.planNewFrame}. */
export type PlanNewFrameResult = {
  scratch: FrameScratch;
  ixCreate: TransactionInstruction;
  /** Frame PDA (same as `scratch.frame`). */
  frame: PublicKey;
  frameBump: number;
};

/**
 * Off-chain mirror of Frame scratch tape: `cursor` + binding indices for `ifx_let`.
 * Plan values with `let*`; build frame instructions via `ix*` (reset / let / assert / patched_cpi / if_else).
 */
export class FrameScratch {
  cursor: number;
  nextIndex: number;
  readonly tapeLen?: number;
  readonly indexCap?: number;
  /** Frame PDA (may be derived before the account exists on-chain). */
  readonly frame: PublicKey;
  /** Ifx program id for all {@link FrameScratch} `ix*` builders (override per call via `IxOpts`). */
  readonly programId: PublicKey;
  private readonly indexTypes = new Map<number, IfxTy>();

  constructor(
    frame: PublicKey,
    tapeLen?: number,
    cursor = 0,
    nextIndex = 0,
    programId: PublicKey = DEFAULT_IFX_PROGRAM_ID
  ) {
    this.frame = frame;
    this.programId = programId;
    this.tapeLen = tapeLen;
    this.indexCap =
      tapeLen === undefined ? undefined : indexCapForTapeLen(tapeLen);
    this.cursor = cursor;
    this.nextIndex = nextIndex;
  }

  /** Sync planner from a fetched frame account (`frame` must match the fetch address). Tests / local debug only. */
  static fromFrame(
    account: DecodedFrame,
    frame: PublicKey,
    programId: PublicKey = DEFAULT_IFX_PROGRAM_ID
  ): FrameScratch {
    return new FrameScratch(
      frame,
      account.tape.length,
      account.cursor,
      account.indexCount,
      programId
    );
  }

  /**
   * Plan a new frame PDA + scratch planner, and the `ifx_create_frame` instruction.
   * `scratch.frame` matches the account `ixCreate` will allocate.
   */
  static planNewFrame(params: CreateIxCreateFrameParams): PlanNewFrameResult {
    const programId = params.programId ?? DEFAULT_IFX_PROGRAM_ID;
    const [frame, frameBump] = framePda(params.payer, params.frameId, programId);
    return {
      scratch: new FrameScratch(frame, params.tapeLen, 0, 0, programId),
      ixCreate: createIxCreateFrame({ ...params, programId }),
      frame,
      frameBump,
    };
  }

  /**
   * Like {@link planNewFrame}, but sets `close_authority` to the Frame PDA itself
   * ({@link immortalCloseAuthority}) — shared / config-pinned Frames with no close Signer.
   * Reset and let remain open to anyone (scratch semantics).
   */
  static planPublicFrame(params: PlanPublicFrameParams): PlanNewFrameResult {
    const programId = params.programId ?? DEFAULT_IFX_PROGRAM_ID;
    return FrameScratch.planNewFrame({
      ...params,
      programId,
      closeAuthority: immortalCloseAuthority(
        params.payer,
        params.frameId,
        programId
      ),
    });
  }

  /** `ifx_create_frame` when you already have a {@link FrameScratch} planner. */
  static ixCreateFrame(
    params: CreateIxCreateFrameParams
  ): TransactionInstruction {
    return createIxCreateFrame(params);
  }

  /** `ifx_close_frame` for this planner's frame PDA. */
  ixCloseFrame(authority: PublicKey, opts?: IxOpts): TransactionInstruction {
    return createIxCloseFrame(this.frame, authority, this.mergeIxOpts(opts));
  }

  private mergeIxOpts(opts?: IxOpts): IxOpts {
    return mergeIxOpts({ programId: this.programId }, opts);
  }

  syncCursor(onChainCursor: number): void {
    this.cursor = onChainCursor;
  }

  syncIndexCount(onChainIndexCount: number): void {
    this.nextIndex = onChainIndexCount;
  }

  /** Fetch and decode this frame account from RPC. Tests / local debug only — not production. */
  async fetchDecodedFrame(
    connection: Connection,
    opts?: FrameScratchReadOpts
  ): Promise<DecodedFrame> {
    const info = await connection.getAccountInfo(
      this.frame,
      opts?.commitment
    );
    if (!info) {
      throw new Error(`frame account missing: ${this.frame.toBase58()}`);
    }
    return decodeFrameAccount(Buffer.from(info.data));
  }

  /** Align local planner with on-chain session from RPC. Tests / local debug only — not production. */
  async refreshFromChain(
    connection: Connection,
    opts?: FrameScratchReadOpts
  ): Promise<DecodedFrame> {
    const decoded = await this.fetchDecodedFrame(connection, opts);
    this.syncCursor(decoded.cursor);
    this.syncIndexCount(decoded.indexCount);
    return decoded;
  }

  /** Next binding index if `ty` were appended now. */
  peekIndex(): number {
    return this.nextIndex;
  }

  /** Start a multi-binding `ifx_let` planner (remaining dedup). */
  letBuilder(): LetIxBuilder {
    return new LetIxBuilder(this);
  }

  letEval<T extends IfxTy>(e: TypedExpr<T>): ScratchValue<T> {
    return this.plan(binding.eval(e));
  }

  letConstU64(n: number | bigint): ScratchValue<"u64"> {
    return this.letEval(expr.u64(n));
  }

  letConstBool(v: boolean): ScratchValue<"bool"> {
    return this.letEval(expr.bool(v));
  }

  letLamports(account: LetAccountInput): ScratchValue<"u64"> {
    const meta = toLetAccountMeta(account);
    return this.plan(binding.accountLamports(0), [meta]);
  }

  /** `remaining[i].data.byteLength` on-chain (`AccountInfo::data_len`). */
  letDataLen(account: LetAccountInput): ScratchValue<"u32"> {
    const meta = toLetAccountMeta(account);
    return this.plan(binding.accountDataLen(0), [meta]);
  }

  letAccountDataSlice<T extends IfxTy>(
    account: LetAccountInput,
    expectedOwner: LetAccountInput,
    ty: T,
    dataOffset: number
  ): ScratchValue<T> {
    const dataMeta = toLetAccountMeta(account);
    const ownerMeta = toLetAccountMeta(expectedOwner);
    return this.plan(
      accountDataSliceBinding(ty, 0, dataOffset, 1),
      [dataMeta, ownerMeta]
    );
  }

  /** `Clock::get()?.slot` (syscall; no remaining account). */
  clockSlot(): ScratchValue<"u64"> {
    return this.plan(binding.sysvarClockSlot());
  }

  clockEpochStartTimestamp(): ScratchValue<"i64"> {
    return this.plan(binding.sysvarClockEpochStartTimestamp());
  }

  clockEpoch(): ScratchValue<"u64"> {
    return this.plan(binding.sysvarClockEpoch());
  }

  clockLeaderScheduleEpoch(): ScratchValue<"u64"> {
    return this.plan(binding.sysvarClockLeaderScheduleEpoch());
  }

  clockUnixTimestamp(): ScratchValue<"i64"> {
    return this.plan(binding.sysvarClockUnixTimestamp());
  }

  /** `Rent::get()?.minimum_balance(dataLen)` — e.g. `165` for a classic SPL token account. */
  rentMinimumBalance(dataLen: number): ScratchValue<"u64"> {
    return this.plan(binding.sysvarRentMinimumBalance(dataLen));
  }

  /** Token-2022 token account (`spl_token_2022` owner). */
  letSplToken2022Amount(account: LetAccountInput): ScratchValue<"u64"> {
    return this.plan(binding.splToken2022AccountAmount(0), [
      toLetAccountMeta(account),
    ]);
  }

  letSplToken2022DelegatedAmount(account: LetAccountInput): ScratchValue<"u64"> {
    return this.plan(binding.splToken2022AccountDelegatedAmount(0), [
      toLetAccountMeta(account),
    ]);
  }

  letSplToken2022AccountState(account: LetAccountInput): ScratchValue<"u8"> {
    return this.plan(binding.splToken2022AccountState(0), [
      toLetAccountMeta(account),
    ]);
  }

  letSplToken2022TransferFeeWithheld(
    account: LetAccountInput
  ): ScratchValue<"u64"> {
    return this.plan(binding.splToken2022AccountTransferFeeWithheld(0), [
      toLetAccountMeta(account),
    ]);
  }

  /** Token-2022 mint (`spl_token_2022` owner). */
  letSplToken2022MintSupply(mint: LetAccountInput): ScratchValue<"u64"> {
    return this.plan(binding.splToken2022MintSupply(0), [
      toLetAccountMeta(mint),
    ]);
  }

  letSplToken2022MintDecimals(mint: LetAccountInput): ScratchValue<"u8"> {
    return this.plan(binding.splToken2022MintDecimals(0), [
      toLetAccountMeta(mint),
    ]);
  }

  letSplToken2022MintTransferFeeBasisPoints(
    mint: LetAccountInput
  ): ScratchValue<"u16"> {
    return this.plan(binding.splToken2022MintTransferFeeBasisPoints(0), [
      toLetAccountMeta(mint),
    ]);
  }

  letSplToken2022MintTransferFeeMaximum(
    mint: LetAccountInput
  ): ScratchValue<"u64"> {
    return this.plan(binding.splToken2022MintTransferFeeMaximum(0), [
      toLetAccountMeta(mint),
    ]);
  }

  letSplToken2022MintWithheldAmount(mint: LetAccountInput): ScratchValue<"u64"> {
    return this.plan(binding.splToken2022MintWithheldAmount(0), [
      toLetAccountMeta(mint),
    ]);
  }

  letSplToken2022MintDefaultAccountState(
    mint: LetAccountInput
  ): ScratchValue<"u8"> {
    return this.plan(binding.splToken2022MintDefaultAccountState(0), [
      toLetAccountMeta(mint),
    ]);
  }

  /**
   * `ifx_reset_frame` — clears on-chain scratch and the local planner.
   * Call once per business tx, then plan `let*` / CPI (same tx). Reuse the returned ix in `tx.add` / `sendAndConfirm`.
   */
  ixReset(opts?: IxOpts): TransactionInstruction {
    this.cursor = 0;
    this.nextIndex = 0;
    this.indexTypes.clear();
    return buildIxResetFrame(this.frame, this.mergeIxOpts(opts));
  }

  /** One `ifx_let` from a single {@link ScratchValue} or a {@link LetIxBuilder} batch. */
  ixLet(
    target: ScratchValue<IfxTy> | LetIxBuilder,
    opts?: IxOpts
  ): TransactionInstruction {
    const ixOpts = this.mergeIxOpts(opts);
    if (target instanceof LetIxBuilder) {
      const { args, remaining } = target.finish();
      return buildIxLet(this.frame, args, remaining, ixOpts);
    }
    return buildIxLet(
      this.frame,
      FrameScratch.toLetArgs([target]),
      [...(target.letRemaining ?? [])],
      ixOpts
    );
  }

  ixAssert(cond: Cond, opts?: IxOpts): TransactionInstruction {
    return buildIxAssert(this.frame, cond, this.mergeIxOpts(opts));
  }

  /** Patched CPI (`ifx_patched_cpi`). */
  ixCpi(built: CpiBuildResult, opts?: IxOpts): TransactionInstruction {
    return createIxCpi(this.frame, built, this.mergeIxOpts(opts));
  }

  /** Conditional CPI arms (`ifx_if_else`). */
  ixIfElse(
    args: IfElseArgs,
    remainingAccounts?: AccountMeta[] | PublicKey[],
    opts?: IxOpts
  ): TransactionInstruction {
    return createIxIfElse(
      this.frame,
      args,
      remainingAccounts,
      this.mergeIxOpts(opts)
    );
  }

  /** @internal */
  plan<T extends IfxTy = IfxTy>(
    letBinding: LetBinding,
    letRemaining?: readonly AccountMeta[]
  ): ScratchValue<T> {
    const ty = inferBindingTy(letBinding, this.indexTypes) as T;
    const valueType = tyForIfxTy(ty);
    const bindingIndex = this.nextIndex;
    if (this.indexCap !== undefined && bindingIndex >= this.indexCap) {
      throw new Error(
        `scratch binding index cap reached (${bindingIndex} >= ${this.indexCap}); use a larger frame tape`
      );
    }
    const { endCursor } = planRecordOffsets(this.cursor, valueType);
    if (this.tapeLen !== undefined && endCursor > this.tapeLen) {
      throw new Error(
        `scratch would exceed tape (${endCursor} > ${this.tapeLen}); reset or use a larger frame (+${recordByteLength(valueType)} B per binding)`
      );
    }
    this.cursor = endCursor;
    this.nextIndex += 1;
    this.indexTypes.set(bindingIndex, ty);
    return scratchValue<T>(
      letBinding,
      { index: bindingIndex },
      letRemaining,
      ty
    );
  }

  /** @internal Used by {@link LetIxBuilder} and SPL helpers. */
  planAtRemainingIndex<T extends IfxTy = IfxTy>(
    letBinding: LetBinding,
    remainingAccountIndex: number
  ): ScratchValue<T> {
    return this.plan(
      remapBindingAccountIndex(letBinding, remainingAccountIndex)
    );
  }

  static toLetArgs(scratchValues: ScratchValue<IfxTy>[]): LetArgs {
    return { bindings: scratchValues.map((p) => p.binding) };
  }
}

// Re-export SPL helpers (use with letBuilder account indices).

import {
  bindSplMintDecimals as letSplMintDecimalsImpl,
  bindSplMintSupply as letSplMintSupplyImpl,
  bindSplTokenAccountState as letSplTokenAccountStateImpl,
  bindSplTokenAmount as letSplTokenAmountImpl,
  bindSplTokenDelegatedAmount as letSplTokenDelegatedAmountImpl,
} from "./spl/bind";

export function letSplTokenAmount(
  scratch: FrameScratch,
  remainingAccountIndex: number
): ScratchValue<"u64"> {
  return letSplTokenAmountImpl(scratch, remainingAccountIndex);
}

export function letSplTokenDelegatedAmount(
  scratch: FrameScratch,
  remainingAccountIndex: number
): ScratchValue<"u64"> {
  return letSplTokenDelegatedAmountImpl(scratch, remainingAccountIndex);
}

export function letSplTokenAccountState(
  scratch: FrameScratch,
  remainingAccountIndex: number
): ScratchValue<"u8"> {
  return letSplTokenAccountStateImpl(scratch, remainingAccountIndex);
}

export function letSplMintSupply(
  scratch: FrameScratch,
  remainingAccountIndex: number
): ScratchValue<"u64"> {
  return letSplMintSupplyImpl(scratch, remainingAccountIndex);
}

export function letSplMintDecimals(
  scratch: FrameScratch,
  remainingAccountIndex: number
): ScratchValue<"u8"> {
  return letSplMintDecimalsImpl(scratch, remainingAccountIndex);
}

import {
  bindSplToken2022AccountAmount,
  bindSplToken2022AccountDelegatedAmount,
  bindSplToken2022AccountState,
  bindSplToken2022AccountTransferFeeWithheld,
  bindSplToken2022MintDecimals,
  bindSplToken2022MintDefaultAccountState,
  bindSplToken2022MintSupply,
  bindSplToken2022MintTransferFeeBasisPoints,
  bindSplToken2022MintTransferFeeMaximum,
  bindSplToken2022MintWithheldAmount,
} from "./spl/token2022-bind";

import {
  bindClockEpoch,
  bindClockEpochStartTimestamp,
  bindClockLeaderScheduleEpoch,
  bindClockSlot,
  bindClockUnixTimestamp,
  bindRentMinimumBalance,
} from "./sysvar/bind";

/** Sysvar reads — `remaining` index not used (syscall). */
export function letClockSlot(scratch: FrameScratch): ScratchValue<"u64"> {
  return bindClockSlot(scratch);
}

export function letClockEpochStartTimestamp(
  scratch: FrameScratch
): ScratchValue<"i64"> {
  return bindClockEpochStartTimestamp(scratch);
}

export function letClockEpoch(scratch: FrameScratch): ScratchValue<"u64"> {
  return bindClockEpoch(scratch);
}

export function letClockLeaderScheduleEpoch(
  scratch: FrameScratch
): ScratchValue<"u64"> {
  return bindClockLeaderScheduleEpoch(scratch);
}

export function letClockUnixTimestamp(scratch: FrameScratch): ScratchValue<"i64"> {
  return bindClockUnixTimestamp(scratch);
}

export function letRentMinimumBalance(
  scratch: FrameScratch,
  dataLen: number
): ScratchValue<"u64"> {
  return bindRentMinimumBalance(scratch, dataLen);
}

/** Token-2022 loads by `remaining_accounts` index (multi-binding batches). */
export function letSplToken2022Amount(
  scratch: FrameScratch,
  remainingAccountIndex: number
): ScratchValue<"u64"> {
  return bindSplToken2022AccountAmount(scratch, remainingAccountIndex);
}

export function letSplToken2022DelegatedAmount(
  scratch: FrameScratch,
  remainingAccountIndex: number
): ScratchValue<"u64"> {
  return bindSplToken2022AccountDelegatedAmount(scratch, remainingAccountIndex);
}

export function letSplToken2022AccountState(
  scratch: FrameScratch,
  remainingAccountIndex: number
): ScratchValue<"u8"> {
  return bindSplToken2022AccountState(scratch, remainingAccountIndex);
}

export function letSplToken2022TransferFeeWithheld(
  scratch: FrameScratch,
  remainingAccountIndex: number
): ScratchValue<"u64"> {
  return bindSplToken2022AccountTransferFeeWithheld(
    scratch,
    remainingAccountIndex
  );
}

export function letSplToken2022MintSupply(
  scratch: FrameScratch,
  remainingAccountIndex: number
): ScratchValue<"u64"> {
  return bindSplToken2022MintSupply(scratch, remainingAccountIndex);
}

export function letSplToken2022MintDecimals(
  scratch: FrameScratch,
  remainingAccountIndex: number
): ScratchValue<"u8"> {
  return bindSplToken2022MintDecimals(scratch, remainingAccountIndex);
}

export function letSplToken2022MintTransferFeeBasisPoints(
  scratch: FrameScratch,
  remainingAccountIndex: number
): ScratchValue<"u16"> {
  return bindSplToken2022MintTransferFeeBasisPoints(
    scratch,
    remainingAccountIndex
  );
}

export function letSplToken2022MintTransferFeeMaximum(
  scratch: FrameScratch,
  remainingAccountIndex: number
): ScratchValue<"u64"> {
  return bindSplToken2022MintTransferFeeMaximum(
    scratch,
    remainingAccountIndex
  );
}

export function letSplToken2022MintWithheldAmount(
  scratch: FrameScratch,
  remainingAccountIndex: number
): ScratchValue<"u64"> {
  return bindSplToken2022MintWithheldAmount(scratch, remainingAccountIndex);
}

export function letSplToken2022MintDefaultAccountState(
  scratch: FrameScratch,
  remainingAccountIndex: number
): ScratchValue<"u8"> {
  return bindSplToken2022MintDefaultAccountState(scratch, remainingAccountIndex);
}

File: go-sdk/scratch/scratch.go
// Package scratch provides FrameScratch and LetBuilder (off-chain tape planner).
package scratch

import (
	"fmt"

	"github.com/gagliardetto/solana-go"
	"github.com/ifx-run/ifx/go-sdk/binding"
	"github.com/ifx-run/ifx/go-sdk/codec"
	"github.com/ifx-run/ifx/go-sdk/constants"
	"github.com/ifx-run/ifx/go-sdk/expr"
	"github.com/ifx-run/ifx/go-sdk/frame"
	"github.com/ifx-run/ifx/go-sdk/frameauthority"
	"github.com/ifx-run/ifx/go-sdk/ix"
	"github.com/ifx-run/ifx/go-sdk/typed"
)

// FrameScratch mirrors on-chain Frame session planning.
type FrameScratch struct {
	Frame     solana.PublicKey
	ProgramID solana.PublicKey
	Authority solana.PublicKey
	TapeLen   *int
	IndexCap  *int
	Cursor    uint32
	NextIndex uint8
	indexTypes map[uint8]typed.IfxTy
}

// PlanNewFrameResult is returned by PlanNewFrame.
type PlanNewFrameResult struct {
	Scratch   *FrameScratch
	IxCreate  solana.Instruction
	Frame     solana.PublicKey
	FrameBump uint8
}

// NewFrameScratch creates a planner for an existing or future Frame PDA.
func NewFrameScratch(framePK solana.PublicKey, tapeLen *int, programID, authority solana.PublicKey) *FrameScratch {
	if programID.IsZero() {
		programID = constants.DefaultProgramID
	}
	s := &FrameScratch{
		Frame:      framePK,
		ProgramID:  programID,
		Authority:  authority,
		TapeLen:    tapeLen,
		indexTypes: make(map[uint8]typed.IfxTy),
	}
	if tapeLen != nil {
		cap, _ := constants.IndexCapForTapeLen(*tapeLen)
		s.IndexCap = &cap
	}
	return s
}

// FromDecodedFrame builds a planner from a decoded account snapshot.
// Tests and local debugging only — not for production.
func FromDecodedFrame(decoded *frame.DecodedFrame, framePK solana.PublicKey, programID solana.PublicKey) *FrameScratch {
	tapeLen := len(decoded.Tape)
	s := NewFrameScratch(framePK, &tapeLen, programID, decoded.Authority)
	s.Cursor = decoded.Cursor
	s.NextIndex = uint8(decoded.IndexCount)
	return s
}

// PlanNewFrameParams configures ifx_create_frame + scratch.
type PlanNewFrameParams struct {
	Payer          solana.PublicKey
	FrameID        [32]byte
	Authority solana.PublicKey
	TapeLen        uint32
	ProgramID      solana.PublicKey
}

// PlanNewFrame returns scratch + create ix for a new Frame PDA.
func PlanNewFrame(p PlanNewFrameParams) (PlanNewFrameResult, error) {
	programID := p.ProgramID
	if programID.IsZero() {
		programID = constants.DefaultProgramID
	}
	framePK, bump, err := frame.FramePDA(programID, p.Payer, p.FrameID)
	if err != nil {
		return PlanNewFrameResult{}, err
	}
	tapeLen := int(p.TapeLen)
	scratch := NewFrameScratch(framePK, &tapeLen, programID, p.Authority)
	ixCreate, _, err := ix.BuildCreateFrame(ix.CreateFrameParams{
		Payer:     p.Payer,
		FrameID:   p.FrameID,
		Authority: p.Authority,
		TapeLen:        p.TapeLen,
		Options:        &ix.Options{ProgramID: programID},
	})
	if err != nil {
		return PlanNewFrameResult{}, err
	}
	return PlanNewFrameResult{
		Scratch:   scratch,
		IxCreate:  ixCreate,
		Frame:     framePK,
		FrameBump: bump,
	}, nil
}

// PlanPublicFrame sets authority to the Frame PDA (public Frame).
func PlanPublicFrame(p PlanNewFrameParams) (PlanNewFrameResult, error) {
	programID := p.ProgramID
	if programID.IsZero() {
		programID = constants.DefaultProgramID
	}
	p.Authority = frameauthority.PublicFrameAuthority(p.Payer, p.FrameID, programID)
	return PlanNewFrame(p)
}

// ForPublicFrame returns a planner for an existing public Frame (authority == frame PDA).
// Use PlanPublicFrame once to create the account; then ForPublicFrame for each business tx.
func ForPublicFrame(framePK, programID solana.PublicKey, tapeLen *int) *FrameScratch {
	if programID.IsZero() {
		programID = constants.DefaultProgramID
	}
	return NewFrameScratch(framePK, tapeLen, programID, framePK)
}

// LetBuilder starts a multi-binding ifx_let batch.
func (s *FrameScratch) LetBuilder() *LetBuilder {
	return &LetBuilder{
		scratch:    s,
		indexByKey: make(map[string]int),
	}
}

func (s *FrameScratch) ixOpts() *ix.Options {
	return &ix.Options{ProgramID: s.ProgramID}
}

// IxReset returns ifx_reset_frame and clears local planner state.
func (s *FrameScratch) IxReset() solana.Instruction {
	s.Cursor = 0
	s.NextIndex = 0
	s.indexTypes = make(map[uint8]typed.IfxTy)
	return ix.BuildResetFrame(s.Frame, s.Authority, s.ixOpts())
}

// IxLet emits ifx_let for one ScratchValue or a LetBuilder batch.
func (s *FrameScratch) IxLet(target interface{}) (solana.Instruction, error) {
	switch v := target.(type) {
	case *LetBuilder:
		fin := v.Finish()
		return ix.BuildLet(s.Frame, s.Authority, fin.Args, fin.RemainingPubkeys(), s.ixOpts())
	case typed.ScratchValue:
		args := codec.LetArgs{Bindings: []binding.Node{v.Binding}}
		var rem []solana.PublicKey
		for _, m := range v.Remaining {
			rem = append(rem, solana.MustPublicKeyFromBase58(m.Pubkey))
		}
		return ix.BuildLet(s.Frame, s.Authority, args, rem, s.ixOpts())
	default:
		return nil, fmt.Errorf("IxLet expects typed.ScratchValue or *LetBuilder")
	}
}

// IxAssert returns ifx_assert.
func (s *FrameScratch) IxAssert(cond interface{}) (solana.Instruction, error) {
	c, err := ToCond(cond)
	if err != nil {
		return nil, err
	}
	return ix.BuildAssert(s.Frame, c, s.ixOpts())
}

// IxAssertMulti returns ifx_assert_multi.
func (s *FrameScratch) IxAssertMulti(conds []interface{}) (solana.Instruction, error) {
	nodes := make([]expr.Node, len(conds))
	for i, c := range conds {
		node, err := ToCond(c)
		if err != nil {
			return nil, err
		}
		nodes[i] = node
	}
	return ix.BuildAssertMulti(s.Frame, nodes, s.ixOpts())
}

// IxCpi returns ifx_patched_cpi.
func (s *FrameScratch) IxCpi(built codec.WireBuildResult) (solana.Instruction, error) {
	return ix.BuildCpi(s.Frame, built, s.ixOpts())
}

// IxIfElse returns ifx_if_else.
func (s *FrameScratch) IxIfElse(args codec.IfElseArgs, remaining []typed.AccountMeta) (solana.Instruction, error) {
	return ix.BuildIfElse(s.Frame, args, ToSolanaMetas(remaining), s.ixOpts())
}

// IxCloseFrame returns ifx_close_frame.
func (s *FrameScratch) IxCloseFrame(authority solana.PublicKey) solana.Instruction {
	return ix.BuildCloseFrame(s.Frame, authority, s.ixOpts())
}

// LetEval plans an Eval binding.
func (s *FrameScratch) LetEval(e expr.Node) (typed.ScratchValue, error) {
	return s.plan(binding.EvalExpr(e), nil)
}

func (s *FrameScratch) LetConstU64(n uint64) (typed.ScratchValue, error) {
	return s.LetEval(expr.U64(n))
}

func (s *FrameScratch) LetConstBool(v bool) (typed.ScratchValue, error) {
	return s.LetEval(expr.Bool(v))
}

func (s *FrameScratch) LetLamports(account interface{}) (typed.ScratchValue, error) {
	meta := toLetMeta(account)
	return s.plan(binding.AccountLamports(0), []typed.AccountMeta{meta})
}

func (s *FrameScratch) LetDataLen(account interface{}) (typed.ScratchValue, error) {
	meta := toLetMeta(account)
	return s.plan(binding.AccountDataLen(0), []typed.AccountMeta{meta})
}

func (s *FrameScratch) LetAccountDataSlice(account, expectedOwner interface{}, ty typed.IfxTy, offset uint32) (typed.ScratchValue, error) {
	tag, err := typed.ValueTypeTag(ty)
	if err != nil {
		return typed.ScratchValue{}, err
	}
	dataMeta := toLetMeta(account)
	ownerMeta := toLetMeta(expectedOwner)
	return s.plan(binding.AccountDataSlice{
		ValueTypeTag:         tag,
		AccountIndex:         0,
		Offset:               offset,
		ExpectedProgramOwner: 1,
	}, []typed.AccountMeta{dataMeta, ownerMeta})
}

func (s *FrameScratch) ClockSlot() (typed.ScratchValue, error) {
	return s.plan(binding.SysvarClockSlot(), nil)
}

func (s *FrameScratch) ClockUnixTimestamp() (typed.ScratchValue, error) {
	return s.plan(binding.SysvarClockUnixTimestamp(), nil)
}

func (s *FrameScratch) RentMinimumBalance(dataLen uint32) (typed.ScratchValue, error) {
	return s.plan(binding.SysvarRentMinimumBalance(dataLen), nil)
}

func (s *FrameScratch) SplTokenAmount(account interface{}) (typed.ScratchValue, error) {
	meta := toLetMeta(account)
	return s.plan(binding.SplTokenAccountAmount(0), []typed.AccountMeta{meta})
}

func (s *FrameScratch) SplMintDecimals(account interface{}) (typed.ScratchValue, error) {
	meta := toLetMeta(account)
	return s.plan(binding.SplMintDecimals(0), []typed.AccountMeta{meta})
}

// PlanAtRemainingIndex plans a binding with a fixed remaining account index (LetBuilder internal).
func (s *FrameScratch) PlanAtRemainingIndex(b binding.Node, accountIndex uint8) (typed.ScratchValue, error) {
	return s.plan(binding.RemapAccountIndex(b, accountIndex), nil)
}

func (s *FrameScratch) plan(b binding.Node, letRemaining []typed.AccountMeta) (typed.ScratchValue, error) {
	ty, err := typed.InferBindingTy(b, s.indexTypes)
	if err != nil {
		return typed.ScratchValue{}, err
	}
	tag, err := typed.ValueTypeTag(ty)
	if err != nil {
		return typed.ScratchValue{}, err
	}
	bindingIndex := s.NextIndex
	if s.IndexCap != nil && int(bindingIndex) >= *s.IndexCap {
		return typed.ScratchValue{}, fmt.Errorf(
			"scratch binding index cap reached (binding #%d >= %d); %d bindings planned — increase tapeLen or reset",
			bindingIndex, *s.IndexCap, s.NextIndex,
		)
	}
	_, _, endCursor, err := frame.PlanRecordOffsets(s.Cursor, tag)
	if err != nil {
		return typed.ScratchValue{}, err
	}
	if s.TapeLen != nil && int(endCursor) > *s.TapeLen {
		recLen, _ := frame.RecordByteLength(tag)
		return typed.ScratchValue{}, fmt.Errorf(
			"scratch would exceed tape (cursor %d + %d B → %d > tapeLen %d, binding #%d); %d bindings planned — reset or use a larger Frame",
			s.Cursor, recLen, endCursor, *s.TapeLen, bindingIndex, s.NextIndex,
		)
	}
	s.Cursor = endCursor
	s.NextIndex++
	s.indexTypes[bindingIndex] = ty
	return typed.ScratchValue{
		Binding:   b,
		Index:     bindingIndex,
		Ty:        ty,
		Remaining: letRemaining,
	}, nil
}

// ToLetArgs collects bindings from scratch values.
func ToLetArgs(values []typed.ScratchValue) codec.LetArgs {
	bindings := make([]binding.Node, len(values))
	for i, v := range values {
		bindings[i] = v.Binding
	}
	return codec.LetArgs{Bindings: bindings}
}
